metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -174,15 +175,27 @@ ControllerResult<Map<String, ApiError>> updateFeatures(
         Map<String, FeatureUpdate.UpgradeType> upgradeTypes,
         boolean validateOnly
     ) {
+        boolean updateFailed = false;
         TreeMap<String, ApiError> results = new TreeMap<>();
         List<ApiMessageAndVersion> records =
                 BoundedList.newArrayBacked(MAX_RECORDS_PER_USER_OP);
+
+        Map<String, Short> proposedUpdatedVersions = new HashMap<>();
+        finalizedVersions.forEach(proposedUpdatedVersions::put);
+        proposedUpdatedVersions.put(MetadataVersion.FEATURE_NAME, metadataVersion.get().featureLevel());
+        updates.forEach(proposedUpdatedVersions::put);
+
         for (Entry<String, Short> entry : updates.entrySet()) {
-            results.put(entry.getKey(), updateFeature(entry.getKey(), entry.getValue(),
-                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records));
+            ApiError error = updateFeature(entry.getKey(), entry.getValue(),
+                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records, proposedUpdatedVersions);
+            results.put(entry.getKey(), error);
+            if (!error.error().equals(Errors.NONE)) {
+                updateFailed = true;
+                break;
+            }
         }
 
-        if (validateOnly) {
+        if (validateOnly || updateFailed) {
```

Comment:
This works too. I can do that.

Reply:
Taking a closer look at this -- are you suggesting to return a different type of controller response here? There are a few options. One is to simply return the feature and error that caused the failure above (in the if block). Another is to modify the type to account for top level errors.

Reply:
I was thinking of just having the following schema. We can return the cause in the top level error message.

```
{
  "apiKey": 57,
  "type": "response",
  "name": "UpdateFeaturesResponse",
  "validVersions": "0-1",
  "flexibleVersions": "0+",
  "fields": [
    { "name": "ThrottleTimeMs", "type": "int32", "versions": "0+",
      "about": "The duration in milliseconds for which the request was throttled due to a quota violation, or zero if the request did not violate any quota." },
    { "name": "ErrorCode", "type": "int16", "versions": "0+",
      "about": "The top-level error code, or `0` if there was no top-level error." },
    { "name": "ErrorMessage", "type": "string", "versions": "0+", "nullableVersions": "0+",
      "about": "The top-level error message, or `null` if there was no top-level error." }
  ]
}
```

Reply:
Right. This makes sense. I think the part I was not sure about was the return type of ControllerResult. I think the approach I am taking right now is to return ControllerResult with a single entry containing the feature with the error. In QuorumController, I convert this to the top level error to return. 

Ie.

```
            if (!error.error().equals(Errors.NONE)) {
                return ControllerResult.of(Collections.emptyList(), Collections.singletonMap(entry.getKey(), error));
            }
```

```
            if (errorEntry.isPresent()) {
                String errorFeatureName = errorEntry.get().getKey();
                ApiError topError = errorEntry.get().getValue();
                String errorString = errorFeatureName + ":" + topError.error().exceptionName() + " (" + topError.message() + ")";
                responseData.setErrorCode(topError.error().code());
                responseData.setErrorMessage("The update failed for all features since the following feature had an error: " + errorString);
 ```   

thinking about ways to clean up how we find the one error. Will update again after lunch.

Reply:
Got it. Maybe ControllerResult just need to include a single error code and error message.

Reply:
I've pushed the new approach. Let me know if it makes sense. (I can also clean up the surrounding code if we want to do it this way. 👍